### PR TITLE
Fix nested recursion

### PR DESCRIPTION
**Pull request by robz**
When recursive functions are created inside if statements or loops, it causes several bugs in the Hack output.

If `--enable vardecl` is passed, the functions are not put into Ref->contents. If `--disable vardecl` is passed, then they are set to Ref->contents, but the Ref itself is never created.

These commits add tests that show off the bugs, and then fix the bugs by using the logic that detects when to use `->contents` in variable declarations for adding Refs. I also moved vardecl logic into php_from_rehp, since it's only needed for Hack, and because it's convenient to include it with the Ref creation logic.

### CODE REVIEW

**Comment by hhugo** [comment]
I'm having hard times understand the issue this PRs is trying to solve (The diff is huge). Would you mind giving a summary in the PR description ?

**Comment by robz** [comment]
@hhugo sorry about that, I added a summary, and the intermediate commit messages are more detailed. The large size of the PR is mostly because I refactored php_from_rehp while tracking down the issues.

File: rehack_tests/output/recursion/Recursion-novardecl-inline.cma.php/Recursion.php [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    $nobug2 = (dynamic $x) : dynamic ==> {
+      $sub = new Ref();
+      $sub->contents = (dynamic $x) : dynamic ==> {
+        return (int) (1 + $sub->contents($x));
+      };
+      return $sub->contents($x);
+    };
+    $bug = (dynamic $x) : dynamic ==> {
+      if (0 === $x) {
+        $sub->contents = (dynamic $x) : dynamic ==> {
```
Comment (robz) [inline rehack_tests/output/recursion/Recursion-novardecl-inline.cma.php/Recursion.php:32]: bug (sub is never declared)

File: rehack_tests/output/recursion/Recursion-novardecl-inline.cma.php/Recursion.php [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+          return (int) (1 + $sub->contents($x));
+        };
+        return $sub->contents($x);
+      }
+      return 0;
+    };
+    $bug__0 = (dynamic $x) : dynamic ==> {
+      if (0 === $x) {
+        $sub->contents = (dynamic $x) : dynamic ==> {
```
Comment (robz) [inline rehack_tests/output/recursion/Recursion-novardecl-inline.cma.php/Recursion.php:42]: bug (sub is never declared)

File: rehack_tests/output/recursion/Recursion-novardecl-inline.cma.php/Recursion.php [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+        return $sub->contents($x);
+      }
+      return 0;
+    };
+    $M = Vector{0, $bug__0} as dynamic;
+    $bug2 = (dynamic $param) : dynamic ==> {
+      $k = Vector{0, 0} as dynamic;
+      $x = 0 as dynamic;
+      for (;;) {
+        $sub->contents = (dynamic $x) : dynamic ==> {
```
Comment (robz) [inline rehack_tests/output/recursion/Recursion-novardecl-inline.cma.php/Recursion.php:55]: bug (sub is never declared)

File: rehack_tests/output/recursion/Recursion-novardecl-inline.cma.php/Recursion.php [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+        if (10 !== $x) {$x = $d_;continue;}
+        return $k;
+      }
+    };
+    $bug3 = (dynamic $param) : dynamic ==> {
+      $k = Vector{0, 0} as dynamic;
+      $x = 0 as dynamic;
+      for (;;) {
+        if (0 === $x) {
+          $sub->contents = (dynamic $x) : dynamic ==> {
```
Comment (robz) [inline rehack_tests/output/recursion/Recursion-novardecl-inline.cma.php/Recursion.php:71]: bug (sub is never declared)

File: rehack_tests/output/recursion/Recursion-novardecl-noinline.cma.php/Recursion.php [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    $nobug2 = (dynamic $x) : dynamic ==> {
+      $sub = new Ref();
+      $sub->contents = (dynamic $x) : dynamic ==> {
+        return (int) (1 + $sub->contents($x));
+      };
+      return $sub->contents($x);
+    };
+    $bug = (dynamic $x) : dynamic ==> {
+      if (0 === $x) {
+        $sub->contents = (dynamic $x) : dynamic ==> {
```
Comment (robz) [inline rehack_tests/output/recursion/Recursion-novardecl-noinline.cma.php/Recursion.php:32]: bug (sub is never declared)

File: rehack_tests/output/recursion/Recursion-novardecl-noinline.cma.php/Recursion.php [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+    };
+    $bug__0 = (dynamic $x) : dynamic ==> {
+      return 0 === $x ? $bug_helper($x) : (0);
+    };
+    $M = Vector{0, $bug__0} as dynamic;
+    $bug2 = (dynamic $param) : dynamic ==> {
+      $k = Vector{0, 0} as dynamic;
+      $x = 0 as dynamic;
+      for (;;) {
+        $sub->contents = (dynamic $x) : dynamic ==> {
```
Comment (robz) [inline rehack_tests/output/recursion/Recursion-novardecl-noinline.cma.php/Recursion.php:55]: bug (sub is never declared)

File: rehack_tests/output/recursion/Recursion-novardecl-noinline.cma.php/Recursion.php [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+        if (10 !== $x) {$x = $d_;continue;}
+        return $k;
+      }
+    };
+    $bug3 = (dynamic $param) : dynamic ==> {
+      $k = Vector{0, 0} as dynamic;
+      $x = 0 as dynamic;
+      for (;;) {
+        if (0 === $x) {
+          $sub->contents = (dynamic $x) : dynamic ==> {
```
Comment (robz) [inline rehack_tests/output/recursion/Recursion-novardecl-noinline.cma.php/Recursion.php:71]: bug (sub is never declared)

File: rehack_tests/output/recursion/Recursion-vardecl-inline.cma.php/Recursion.php [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+      $sub = new Ref();
+      $sub->contents = (dynamic $x) : dynamic ==> {
+        return (int) (1 + $sub->contents($x));
+      };
+      return $sub->contents($x);
+    };
+    $bug = (dynamic $x) : dynamic ==> {
+      $sub = null as dynamic;
+      if (0 === $x) {
+        $sub = (dynamic $x) : dynamic ==> {return (int) (1 + $sub($x));};
```
Comment (robz) [inline rehack_tests/output/recursion/Recursion-vardecl-inline.cma.php/Recursion.php]: bug (sub should be a Ref)

File: rehack_tests/output/recursion/Recursion-vardecl-inline.cma.php/Recursion.php [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+        return $sub($x);
+      }
+      return 0;
+    };
+    $bug__0 = (dynamic $x) : dynamic ==> {
+      $sub = null as dynamic;
+      if (0 === $x) {
+        $sub = (dynamic $x) : dynamic ==> {return (int) (1 + $sub($x));};
```
Comment (robz) [inline rehack_tests/output/recursion/Recursion-vardecl-inline.cma.php/Recursion.php]: bug (sub should be a Ref)

File: rehack_tests/output/recursion/Recursion-vardecl-inline.cma.php/Recursion.php [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+    };
+    $M = Vector{0, $bug__0} as dynamic;
+    $bug2 = (dynamic $param) : dynamic ==> {
+      $sub = null as dynamic;
+      $c_ = null as dynamic;
+      $d_ = null as dynamic;
+      $k = Vector{0, 0} as dynamic;
+      $x = 0 as dynamic;
+      for (;;) {
+        $sub = (dynamic $x) : dynamic ==> {return (int) (1 + $sub($x));};
```
Comment (robz) [inline rehack_tests/output/recursion/Recursion-vardecl-inline.cma.php/Recursion.php]: bug (sub should be a Ref)

File: rehack_tests/output/recursion/Recursion-vardecl-inline.cma.php/Recursion.php [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+    };
+    $bug3 = (dynamic $param) : dynamic ==> {
+      $sub = null as dynamic;
+      $a_ = null as dynamic;
+      $b_ = null as dynamic;
+      $k = Vector{0, 0} as dynamic;
+      $x = 0 as dynamic;
+      for (;;) {
+        if (0 === $x) {
+          $sub = (dynamic $x) : dynamic ==> {return (int) (1 + $sub($x));};
```
Comment (robz) [inline rehack_tests/output/recursion/Recursion-vardecl-inline.cma.php/Recursion.php]: bug (sub should be a Ref)

File: rehack_tests/output/recursion/Recursion-vardecl-noinline.cma.php/Recursion.php [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+      $sub = new Ref();
+      $sub->contents = (dynamic $x) : dynamic ==> {
+        return (int) (1 + $sub->contents($x));
+      };
+      return $sub->contents($x);
+    };
+    $bug = (dynamic $x) : dynamic ==> {
+      $sub = null as dynamic;
+      if (0 === $x) {
+        $sub = (dynamic $x) : dynamic ==> {return (int) (1 + $sub($x));};
```
Comment (robz) [inline rehack_tests/output/recursion/Recursion-vardecl-noinline.cma.php/Recursion.php]: bug (sub should be a Ref)

File: rehack_tests/output/recursion/Recursion-vardecl-noinline.cma.php/Recursion.php [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+    };
+    $M = Vector{0, $bug__0} as dynamic;
+    $bug2 = (dynamic $param) : dynamic ==> {
+      $sub = null as dynamic;
+      $c_ = null as dynamic;
+      $d_ = null as dynamic;
+      $k = Vector{0, 0} as dynamic;
+      $x = 0 as dynamic;
+      for (;;) {
+        $sub = (dynamic $x) : dynamic ==> {return (int) (1 + $sub($x));};
```
Comment (robz) [inline rehack_tests/output/recursion/Recursion-vardecl-noinline.cma.php/Recursion.php]: bug (sub should be a Ref)

File: rehack_tests/output/recursion/Recursion-vardecl-noinline.cma.php/Recursion.php [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+    };
+    $bug3 = (dynamic $param) : dynamic ==> {
+      $sub = null as dynamic;
+      $a_ = null as dynamic;
+      $b_ = null as dynamic;
+      $k = Vector{0, 0} as dynamic;
+      $x = 0 as dynamic;
+      for (;;) {
+        if (0 === $x) {
+          $sub = (dynamic $x) : dynamic ==> {return (int) (1 + $sub($x));};
```
Comment (robz) [inline rehack_tests/output/recursion/Recursion-vardecl-noinline.cma.php/Recursion.php]: bug (sub should be a Ref)

**Pull request merged**